WordPress/wordpress-develop
# Eliminate manual construction of script tags in WP_Scripts and pass other scripts through wp_print_inline_script_tag()

The scope here is now limited to the frontend (including Customizer preview) as well as the login screen (`wp-login.php`). Admin screens are not included since this would increase the scope significantly. Additionally, the core themes have not been updated since `wp_print_inline_script_tag()` was introduced in WP 5.7, and the last theme to include any custom scripts (Twenty Twenty-One) supports WordPress 5.3 and above. So to take advantage of the example [Strict CSP plugin](https://gist.github.com/westonruter/c8b49406391a8d86a5864fb41a523ae9) with core themes, you have to use one of these themes which don't manually construct scripts:

* Twenty Eleven (only uses manual script tag in IE conditional comment)
* Twenty Twelve (only uses manual script tag in IE conditional comment)
* Twenty Thirteen
* Twenty Fourteen (only uses manual script tag in IE conditional comment)
* Twenty Sixteen
* Twenty Nineteen
* Twenty Twenty-Two (a block theme, so no custom JS)
* Twenty Twenty-Three (a block theme, so no custom JS)





Previously: https://github.com/10up/wordpress-develop/pull/58

See https://github.com/WordPress/gutenberg/pull/54637 for syncing block change to Gutenberg.

# Testing Instructions

1. Install and activate the [Strict CSP plugin](https://gist.github.com/westonruter/c8b49406391a8d86a5864fb41a523ae9)
2. Activate Twenty Twenty-Four
3. Navigate the frontend, the wp-login screen, and the Customizer preview (ideally so all scripts touched end up being printed) with the console open, making sure that there are no CSP warnings like: 
> Refused to load the script 'http://localhost:8889/wp-includes/blocks/navigation/view.min.js?ver=5687b52bb8c84fb4ae68' because it violates the following Content Security Policy directive: "script-src 'nonce-6aa1ba7b65' 'unsafe-inline' 'strict-dynamic' https: http:". Note that 'strict-dynamic' is present, so host-based allowlisting is disabled. Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.

Trac ticket: https://core.trac.wordpress.org/ticket/58664



# Commit Message

Script Loader: Use `wp_get_script_tag()` and `wp_get_inline_script_tag()`/`wp_print_inline_script_tag()` helper functions to output scripts on frontend and login screen.

Using script tag helper functions allows plugins to employ the `wp_script_attributes` and `wp_inline_script_attributes` filters to inject the `nonce` attribute to apply Content Security Policy (e.g. Strict CSP). Use of helper functions also simplifies logic in `WP_Scripts`.

* Update `wp_get_inline_script_tag()` to wrap inline script in CDATA blocks for XHTML-compatibility.
* Ensure the `type` attribute is printed first in `wp_get_inline_script_tag()` for back-compat.
* Wrap existing `<script>` tags in output buffering to retain IDE supports.
* In `wp_get_inline_script_tag()`, append the newline to `$javascript` before it is passed into the `wp_inline_script_attributes` filter so that the CSP hash can be computed properly.
* In `the_block_template_skip_link()`, opt to enqueue the inline script rather than print it.
* Add `ext-php` to `composer.json` under `suggest` as previously it was an undeclared dependency for running PHPUnit tests.
* Update tests to rely on `DOMDocument` to compare script markup, normalizing unsemantic differences.

Props westonruter, spacedmonkey, flixos90, 10upsimon, dmsnell, mukesh27, joemcgill, swissspidy, azaozz.
Fixes #58664.
See #39941.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**
